### PR TITLE
have file_stem accept a full path

### PR DESCRIPTION
**Pull request by mattseddon**
Follow up to https://github.com/iterative/datachain/pull/101/files#r1697409039 and https://github.com/iterative/datachain/pull/101/files#r1709153373

This PR updates the compiled `file_stem` function to accept a full path instead of expecting users to write `path.file_stem(path.name(C("file.path")))`. This change leads to (slightly) better SQL being generated as well.

For comparison I've put the current and newly generated SQL below:

### Before PR

```sql
              rtrim(
                substr(
                  ltrim(
                    substr(
                      source__file__path,
                      length(
                        rtrim(
                          rtrim(
                            source__file__path,
                            replace(source__file__path, :param_1, :param_2)
                          ),
                          :param_1
                        )
                      ) + :length_16
                    ),
                    :param_1
                  ),
                  :substr_2,
                  length(
                    ltrim(
                      substr(
                        source__file__path,
                        length(
                          rtrim(
                            rtrim(
                              source__file__path,
                              replace(source__file__path, :param_1, :param_2)
                            ),
                            :param_1
                          )
                        ) + :length_17
                      ),
                      :param_1
                    )
                  ) - CASE WHEN (
                    instr(
                      ltrim(
                        substr(
                          ltrim(
                            substr(
                              source__file__path,
                              length(
                                rtrim(
                                  rtrim(
                                    source__file__path,
                                    replace(source__file__path, :param_1, :param_2)
                                  ),
                                  :param_1
                                )
                              ) + :length_18
                            ),
                            :param_1
                          ),
                          length(
                            rtrim(
                              rtrim(
                                ltrim(
                                  substr(
                                    source__file__path,
                                    length(
                                      rtrim(
                                        rtrim(
                                          source__file__path,
                                          replace(source__file__path, :param_1, :param_2)
                                        ),
                                        :param_1
                                      )
                                    ) + :length_19
                                  ),
                                  :param_1
                                ),
                                replace(
                                  ltrim(
                                    substr(
                                      source__file__path,
                                      length(
                                        rtrim(
                                          rtrim(
                                            source__file__path,
                                            replace(source__file__path, :param_1, :param_2)
                                          ),
                                          :param_1
                                        )
                                      ) + :length_20
                                    ),
                                    :param_1
                                  ),
                                  :param_1,
                                  :param_2
                                )
                              ),
                              :param_1
                            )
                          ) + :length_21
                        ),
                        :param_1
                      ),
                      :param_3
                    ) = :instr_2
                  ) THEN :param_5 ELSE length(
                    ltrim(
                      substr(
                        ltrim(
                          substr(
                            source__file__path,
                            length(
                              rtrim(
                                rtrim(
                                  source__file__path,
                                  replace(source__file__path, :param_1, :param_2)
                                ),
                                :param_1
                              )
                            ) + :length_22
                          ),
                          :param_1
                        ),
                        length(
                          rtrim(
                            rtrim(
                              ltrim(
                                substr(
                                  source__file__path,
                                  length(
                                    rtrim(
                                      rtrim(
                                        source__file__path,
                                        replace(source__file__path, :param_1, :param_2)
                                      ),
                                      :param_1
                                    )
                                  ) + :length_23
                                ),
                                :param_1
                              ),
                              replace(
                                ltrim(
                                  substr(
                                    source__file__path,
                                    length(
                                      rtrim(
                                        rtrim(
                                          source__file__path,
                                          replace(source__file__path, :param_1, :param_2)
                                        ),
                                        :param_1
                                      )
                                    ) + :length_24
                                  ),
                                  :param_1
                                ),
                                :param_1,
                                :param_2
                              )
                            ),
                            :param_1
                          )
                        ) + :length_21
                      ),
                      :param_1
                    )
                  ) - length(
                    rtrim(
                      ltrim(
                        substr(
                          ltrim(
                            substr(
                              source__file__path,
                              length(
                                rtrim(
                                  rtrim(
                                    source__file__path,
                                    replace(source__file__path, :param_1, :param_2)
                                  ),
                                  :param_1
                                )
                              ) + :length_25
                            ),
                            :param_1
                          ),
                          length(
                            rtrim(
                              rtrim(
                                ltrim(
                                  substr(
                                    source__file__path,
                                    length(
                                      rtrim(
                                        rtrim(
                                          source__file__path,
                                          replace(source__file__path, :param_1, :param_2)
                                        ),
                                        :param_1
                                      )
                                    ) + :length_26
                                  ),
                                  :param_1
                                ),
                                replace(
                                  ltrim(
                                    substr(
                                      source__file__path,
                                      length(
                                        rtrim(
                                          rtrim(
                                            source__file__path,
                                            replace(source__file__path, :param_1, :param_2)
                                          ),
                                          :param_1
                                        )
                                      ) + :length_27
                                    ),
                                    :param_1
                                  ),
                                  :param_1,
                                  :param_2
                                )
                              ),
                              :param_1
                            )
                          ) + :length_21
                        ),
                        :param_1
                      ),
                      replace(
                        ltrim(
                          substr(
                            ltrim(
                              substr(
                                source__file__path,
                                length(
                                  rtrim(
                                    rtrim(
                                      source__file__path,
                                      replace(source__file__path, :param_1, :param_2)
                                    ),
                                    :param_1
                                  )
                                ) + :length_28
                              ),
                              :param_1
                            ),
                            length(
                              rtrim(
                                rtrim(
                                  ltrim(
                                    substr(
                                      source__file__path,
                                      length(
                                        rtrim(
                                          rtrim(
                                            source__file__path,
                                            replace(source__file__path, :param_1, :param_2)
                                          ),
                                          :param_1
                                        )
                                      ) + :length_29
                                    ),
                                    :param_1
                                  ),
                                  replace(
                                    ltrim(
                                      substr(
                                        source__file__path,
                                        length(
                                          rtrim(
                                            rtrim(
                                              source__file__path,
                                              replace(source__file__path, :param_1, :param_2)
                                            ),
                                            :param_1
                                          )
                                        ) + :length_30
                                      ),
                                      :param_1
                                    ),
                                    :param_1,
                                    :param_2
                                  )
                                ),
                                :param_1
                              )
                            ) + :length_21
                          ),
                          :param_1
                        ),
                        :param_3,
                        :param_2
                      )
                    )
                  ) END
                ),
                :param_3
              ) AS stem
```

### After PR

```sql
              CASE WHEN (
                instr(file__path, '/') = 0
              ) THEN rtrim(
                substr(
                  file__path, 
                  1, 
                  length(file__path) - CASE WHEN (
                    instr(file__path, '.') = 0
                  ) THEN 0 ELSE length(file__path) - length(
                    rtrim(
                      file__path, 
                      replace(file__path, '.', '')
                    )
                  ) END
                ), 
                '.'
              ) ELSE ltrim(
                rtrim(
                  substr(
                    file__path, 
                    length(
                      rtrim(
                        rtrim(
                          file__path, 
                          replace(file__path, '/', '')
                        ), 
                        '/'
                      )
                    ) + 1, 
                    (
                      length(file__path) - length(
                        rtrim(
                          rtrim(
                            file__path, 
                            replace(file__path, '/', '')
                          ), 
                          '/'
                        )
                      )
                    ) - CASE WHEN (
                      instr(
                        ltrim(
                          substr(
                            file__path, 
                            length(
                              rtrim(
                                rtrim(
                                  file__path, 
                                  replace(file__path, '/', '')
                                ), 
                                '/'
                              )
                            ) + 1
                          ), 
                          '/'
                        ), 
                        '.'
                      ) = 0
                    ) THEN 0 ELSE length(
                      ltrim(
                        substr(
                          file__path, 
                          length(
                            rtrim(
                              rtrim(
                                file__path, 
                                replace(file__path, '/', '')
                              ), 
                              '/'
                            )
                          ) + 1
                        ), 
                        '/'
                      )
                    ) - length(
                      rtrim(
                        ltrim(
                          substr(
                            file__path, 
                            length(
                              rtrim(
                                rtrim(
                                  file__path, 
                                  replace(file__path, '/', '')
                                ), 
                                '/'
                              )
                            ) + 1
                          ), 
                          '/'
                        ), 
                        replace(
                          ltrim(
                            substr(
                              file__path, 
                              length(
                                rtrim(
                                  rtrim(
                                    file__path, 
                                    replace(file__path, '/', '')
                                  ), 
                                  '/'
                                )
                              ) + 1
                            ), 
                            '/'
                          ), 
                          '.', 
                          ''
                        )
                      )
                    ) END
                  ), 
                  '.'
                ), 
                '/'
              ) END AS stem, 
```

... as you can see the SQL is still hot garbage but there is less of it and the UX is slightly better

### CODE REVIEW

File: examples/get_started/common_sql_functions.py [hunk 1]
```diff
@@ -26,8 +26,8 @@ def num_chars_udf(file):
 
 (
     dc.mutate(
-        stem=path.file_stem(path.name(C("file.path"))),
-        ext=path.file_ext(path.name(C("file.path"))),
+        stem=path.file_stem(C("file.path")),
+        ext=path.file_ext(C("file.path")),
```
Comment (mattseddon) [inline examples/get_started/common_sql_functions.py:30]: [F] This accepted a full path already

**Comment by mattseddon** [comment]
Will need a studio companion PR.

**Review by dtulga — APPROVED** [review]
LGTM, thanks for adding this! Although, I would recommend only merging once the Studio/ClickHouse part is done and those tests pass as well.

**Comment by mattseddon** [comment]
> LGTM, thanks for adding this! Although, I would recommend only merging once the Studio/ClickHouse part is done and those tests pass as well.

yeah, I'm a bit stuck. The companion is up but the tests are completely borked (see #292 / #293)

**Comment by dtulga** [comment]
Yeah, that is less than ideal - if all the tests pass except those that were already broken, that's fine by me. (Although ideally these tests should be fixed on `main` at some point. 😕 )

**Review by ilongin — APPROVED** [review]

**Review by dreadatour — APPROVED** [review]
Looks good to me! Thank you! 🙏

**Comment by mattseddon** [comment]
blocked by https://github.com/iterative/studio/pull/10469 which is blocked by https://github.com/iterative/studio/pull/10480

**Pull request merged**